File: src/python/121._Best_Time_to_Buy_and_Sell_Stock.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
from typing import List


#  Algo 2 : One-Pass Iteration : O(n) , O(1)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l, r = 0, 1 # l = buy, r = sell
        maxProfit = 0
        
        while l < len(prices) and r < len(prices):
            if prices[l] < prices[r]:
                profit = prices[r] - prices[l]
                maxProfit = max(profit, maxProfit)
            else:
                l = r
            r += 1
        return maxProfit

File: src/java/121._Best_Time_to_Buy_and_Sell_Stock.java
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
// Algo 1 : Brute force : O(n^2) , O(1)
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        for(int i = 0 ; i < prices.length ; i++) {
            for(int j = i+1 ; j < prices.length ; j++) {
                if(maxProfit < (prices[j] - prices[i])) 
                    maxProfit = prices[j] - prices[i];
            }
        }
        return maxProfit;
        
    }
}
// Algo 2 : One-Pass Iteration : O(n) , O(1)
public class Solution {
    public int maxProfit(int prices[]) {
        int maxProfit = Integer.MIN_VALUE , minPrice = Integer.MAX_VALUE;
        for(int i = 0 ; i < prices.length ; i++) {
            minPrice = Math.min(minPrice, prices[i]);
            maxProfit = Math.max(maxProfit, prices[i] - minPrice);
        }
        return maxProfit;
    }
}
